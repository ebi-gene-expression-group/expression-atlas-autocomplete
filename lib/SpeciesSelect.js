'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactRefetch = require('react-refetch');

var _urijs = require('urijs');

var _urijs2 = _interopRequireDefault(_urijs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _option = function _option(label) {
  return _react2.default.createElement(
    'option',
    { key: label, value: label },
    label
  );
};

var SpeciesSelect = function SpeciesSelect(props) {
  var speciesFetch = props.speciesFetch;


  return _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      'label',
      null,
      'Species'
    ),
    speciesFetch.fulfilled ? _react2.default.createElement(
      'select',
      { onChange: props.onChange },
      _react2.default.createElement(
        'option',
        { value: '' },
        'Any'
      ),
      speciesFetch.value.topSpecies.map(_option),
      _react2.default.createElement(
        'option',
        { value: '-', disabled: 'true' },
        speciesFetch.value.separator
      ),
      speciesFetch.value.allSpecies.map(_option)
    ) : speciesFetch.pending ? _react2.default.createElement(
      'select',
      { disabled: 'true' },
      _option('Fetching species\u2026')
    ) : _react2.default.createElement(
      'select',
      { disabled: 'true' },
      _option('Error fetching species')
    )
  );
};

SpeciesSelect.propTypes = {
  atlasUrl: _propTypes2.default.string.isRequired,
  onChange: _propTypes2.default.func.isRequired
};

exports.default = (0, _reactRefetch.connect)(function (props) {
  return {
    speciesFetch: (0, _urijs2.default)('json/suggestions/species', props.atlasUrl).toString()
  };
})(SpeciesSelect);