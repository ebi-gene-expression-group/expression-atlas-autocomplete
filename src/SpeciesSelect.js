import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-refetch'

import URI from 'urijs'

const _option = (label) => {
  return <option key={label} value={label}>{label}</option>
}

const SpeciesSelect = (props) => {
  const {speciesFetch} = props

  return (
    <div>
      <label>Species</label>
      { speciesFetch.fulfilled ?
        <select onChange={props.onChange}>
          <option value={``}>Any</option>
          {speciesFetch.value.topSpecies.map(_option)}
          <option value={`-`} disabled={`true`}>{speciesFetch.value.separator}</option>
          {speciesFetch.value.allSpecies.map(_option)}
        </select> :

        speciesFetch.pending ?
          <select disabled={`true`}>{_option(`Fetching speciesâ€¦`)}</select> :

          <select disabled={`true`}>{_option(`Error fetching species`)}</select>
      }
    </div>
  )
}

SpeciesSelect.propTypes = {
  atlasUrl: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired
}

export default connect((props) => ({
  speciesFetch: URI(`json/suggestions/species`, props.atlasUrl).toString()
}))(SpeciesSelect)
