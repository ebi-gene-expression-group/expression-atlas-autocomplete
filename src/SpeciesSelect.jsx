import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-refetch'

import URI from 'urijs'

class SpeciesSelect extends React.Component {
  constructor(props) {
    super(props)
  }

  _option(label) {
    return <option key={label} value={label}>{label}</option>
  }

  render() {
    const {speciesFetch} = this.props

    if (speciesFetch.fulfilled) {
      return (
        <div>
          <label>Species</label>
          <select onChange={this.props.onChange}>
            <option value={``}>Any</option>
            {speciesFetch.value.topSpecies.map(this._option)}
            <option value={`-`} disabled={`true`}>{speciesFetch.value.separator}</option>
            {speciesFetch.value.allSpecies.map(this._option)}
          </select>
        </div>
      )
    } else if (speciesFetch.pending) {
      return (
        <div>
          <label>Species</label>
          <select>{this._option(`Waitingâ€¦`)}</select>
        </div>
      )
    } else {
      return (
        <div>
          <label>Species</label>
          <select>{this._option(`Error!`)}</select>
        </div>
      )
    }
  }
}

SpeciesSelect.propTypes = {
  atlasUrl: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired
}

export default connect((props) => ({
  speciesFetch: URI(`json/suggestions/species`, props.atlasUrl).toString()
}))(SpeciesSelect)
