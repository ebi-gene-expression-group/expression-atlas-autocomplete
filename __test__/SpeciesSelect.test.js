import React from 'react'
import TestRenderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import {mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import SpeciesSelect from '../src/SpeciesSelect'

Enzyme.configure({ adapter: new Adapter() })

const props = {
  topSpecies: [`Meeseek`],
  allSpecies: [`Zigerion`, `Cromulon`, `Meeseek`, `Gromflomite`],
  statusMessage: ``,
  onChange: () => {},
  selectedValue: ``
}

describe(`SpeciesSelect`, () => {
  test(`displays a status message, and nothing else, if it’s not empty`, () => {
    const statusMessage = `loading`
    const wrapper = mount(<SpeciesSelect {...props} statusMessage={statusMessage}/>)
    expect(wrapper.find(`option`)).toHaveLength(1)
    expect(wrapper.find(`option`).first().text()).toBe(statusMessage)
  })

  test(`shows a separator if at least one top species is provided`, () => {
    const wrapper = mount(<SpeciesSelect {...props} />)
    expect(wrapper.find(`option`)
        .map((rw) => rw.html())
        .map((staticHtml) => staticHtml.match(/value=\"(.*?)\"/)[1]))
      .toContain(`-`)
  })

  test(`doesn’t show a separator if no top species are provided`, () => {
    const wrapper = mount(<SpeciesSelect {...props} topSpecies={[]}/>)
    expect(wrapper.find(`option`)
        .map((rw) => rw.html())
        .map((staticHtml) => staticHtml.match(/value=\"(.*?)\"/)[1]))
      .not.toContain(`-`)
  })

  test(`shows all species plus “Any” at the top`, () => {
    const wrapper = mount(<SpeciesSelect {...props} />)
    expect(wrapper.find(`option`).first().text()).toBe(`Any`)

  })

  test(`matches snapshot`, () => {
    const testRenderer = TestRenderer.create(<SpeciesSelect {...props} />)
    const tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
