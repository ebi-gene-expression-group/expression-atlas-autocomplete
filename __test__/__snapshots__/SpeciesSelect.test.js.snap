// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpeciesSelect matches snapshot 1`] = `
<div>
  <label>
    Species
  </label>
  <select
    onChange={[Function]}
    value=""
  >
    <option
      value=""
    >
      Any
    </option>
    <option
      value="Meeseek"
    >
      Meeseek
    </option>
    <option
      disabled="true"
      value="-"
    >
      ━━━━━━━━━━━━
    </option>
    <option
      value="Zigerion"
    >
      Zigerion
    </option>
    <option
      value="Cromulon"
    >
      Cromulon
    </option>
    <option
      value="Meeseek"
    >
      Meeseek
    </option>
    <option
      value="Gromflomite"
    >
      Gromflomite
    </option>
  </select>
</div>
`;
